README
NAME
    video-generator - Script to generate video.

SYNOPSIS
     video-generator [-d duration] [-f fps] [-h] [-i input_dir] [-s size] [-v] [--version] output_file

DESCRIPTION
     video-generator script uses Video::Generator class to create videos via ffmpeg.

ARGUMENTS
    *       "-d duration"

             Duration.
             Default value is 10000 (10s).

    *       "-f fps"

             Frame rate.
             Default value is 60 frames per second.

    *       "-h"

             Print help.

    *       "-i input_dir"

             Input directory with images.
             Default value is nothing, then generates random images.

    *       "-s size"

             Video size.
             Default value is 1920x1080.

    *       "-v"

             Verbose mode.

    *       "--version"

             Print version of script.

    *       "output_file"

             Output file.
             e.g. foo.mpg
             e.g. foo.mp4

EXAMPLE1
     video-generator

     # Output:
     # Usage: video-generator [-d duration] [-f fps] [-h]
     #         [-i input_dir] [-s size] [-v] [--version] output_file
     #
     #         -d duration     Duration (default value is 10000 (=10s)
     #         -f fps          Frame rate
     #         -h              Print help.
     #         -i input_dir    Input directory with images (default value is nothing).
     #         -s size         Size (default value is 1920x1080).
     #         -v              Verbose mode.
     #         --version       Print version.

EXAMPLE2
     video-generator foo.mpg

     # Output:
     # -nothing-
     # Created foo.mpg file in local directory.

EXAMPLE3
     video-generator -s 100x100 foo.mp4

     # Output:
     # -nothing-
     # Created foo.mp4 file in local directory.

EXAMPLE4
     video-generator -i /data/images/ foo.mpg

     # Output:
     # -nothing-
     # Created foo.mpg file in local directory, generated from images in /data/images.

REPOSITORY
    <https://github.com/michal-josef-spacek/App-Video-Generator>

AUTHOR
    Michal Josef Špaček <mailto:skim@cpan.org>

    <http://skim.cz>

LICENSE AND COPYRIGHT
    © 2015-2021 Michal Josef Špaček

    BSD 2-Clause License

VERSION
    0.10



Changes
0.10

0.09 2021-02-03T00:38:49+01:00
 - Add test for script POD.
 - Fix bugtracker.
 - Update copyright years.

0.08 2020-04-23T13:30:39+02:00
 - Add App::Image::Generator to SEE ALSO section in doc.
 - Fix Makefile.PL fails with no '.' in @INC.
 - Improve LICENSE AND COPYRIGHT section in doc.
 - Improve SYNOPSIS secion in doc.
 - Update Module::Install to 1.19 version.
 - Update author github username.
 - Update copyright years.
 - Update my name to reality.

0.07 2015-11-06T13:01:08+01:00
 - Fix width and height for -i option.
 - Update examples/ directory.

0.06 2015-09-30T17:18:12+02:00
 - Add support for generating of video from image files.
 - Fix number of chars on line (80).
 - Fix EXAMPLE3 in bin/video-generator doc.

0.05 2015-09-18T14:57:36+02:00
 - Add support for duration suffixes min (minute) and h (hour).

0.04 2015-09-16T11:30:18+02:00
 - Add support for ms/s suffix in -d argument.
 - Fix output in EXAMPLE in doc.
 - Fix man page of video-generator script.
 - Move author tests to xt/ directory.
 - Remove obsolete dependency.

0.03 2015-09-15T22:35:35+02:00
 - Add proper exit code.
 - Add test for constructor.
 - Add verbose mode (-v).
 - Move argument processing to run() method.
 - Remove obsolete code from Generator.pm.
 - Sort argument default values.

0.02 2015-08-20T08:51:16+02:00
 - Add examples/ directory.
 - Fix output of example in doc.
 - Fix repository in doc.
 - Fix script description in doc.
 - Improve help.

0.01 2015-08-09T11:30:04+02:00
 - First version.


Generator.pm
package App::Video::Generator;

use strict;
use warnings;

use English;
use Error::Pure qw(err);
use Getopt::Std;
use Image::Select;
use Readonly;
use Video::Generator;

# Constants.
Readonly::Scalar our $EMPTY_STR => q{};

our $VERSION = 0.10;

# Constructor.
sub new {
	my $class = shift;

	# Create object.
	my $self = bless {}, $class;

	# Object.
	return $self;
}

# Run script.
sub run {
	my $self = shift;

	# Process arguments.
	$self->{'_opts'} = {
		'd' => 10000,
		'f' => 60,
		'h' => 0,
		'i' => $EMPTY_STR,
		's' => '1920x1080',
		'v' => 0,
	};
	if (! getopts('d:f:hi:s:v', $self->{'_opts'}) || @ARGV < 1
		|| $self->{'_opts'}->{'h'}) {

		print STDERR "Usage: $0 [-d duration] [-f fps] [-h]\n\t".
			"[-i input_dir] [-s size] [-v] [--version] ".
			"output_file\n\n";
		print STDERR "\t-d duration\tDuration in numeric value or ".
			"with ms/s/min/h suffix\n\t\t\t(default value is ".
			"10000 [=10s]).\n";
		print STDERR "\t-f fps\t\tFrame rate\n";
		print STDERR "\t-h\t\tPrint help.\n";
		print STDERR "\t-i input_dir\tInput directory with ".
			"images (default value is nothing).\n";
		print STDERR "\t-s size\t\tSize (default value is ".
			"1920x1080).\n";
		print STDERR "\t-v\t\tVerbose mode.\n";
		print STDERR "\t--version\tPrint version.\n";
		exit 1;
	}
	$self->{'_output_file'} = $ARGV[0];

	# Check size.
	if ($self->{'_opts'}->{'s'} !~ m/^(\d+)x(\d+)$/ms) {
		err 'Bad size value.', 'Value', $self->{'_opts'}->{'s'};
	}
	$self->{'_width'} = $1;
	$self->{'_height'} = $2;

	# Run.
	eval {
		# Images from directory.
		my $image_generator;
		if ($self->{'_opts'}->{'i'} ne $EMPTY_STR) {
			$image_generator = Image::Select->new(
				'debug' => ($self->{'_opts'}->{'v'} ? 1 : 0),
				'height' => $self->{'_height'},
				'path_to_images' => $self->{'_opts'}->{'i'},
				'width' => $self->{'_width'},
			);
		}

		my $vg = Video::Generator->new(
			'duration' => $self->{'_opts'}->{'d'},
			'fps' => $self->{'_opts'}->{'f'},
			'height' => $self->{'_height'},
			'image_generator' => $image_generator,
			'verbose' => $self->{'_opts'}->{'v'},
			'width' => $self->{'_width'},
		);
		$vg->create($self->{'_output_file'});
	};
	if ($EVAL_ERROR) {
		err 'Cannot create video.';
	}
	return 0;
}

1;

__END__

=pod

=encoding utf8

=head1 NAME

App::Video::Generator - Perl class for video-generator application.

=head1 SYNOPSIS

 use App::Video::Generator;

 my $obj = App::Video::Generator->new;
 $obj->run;

=head1 METHODS

=over 8

=item C<new()>

 Constructor.

=item C<run()>

 Run.

=back

=head1 ERRORS

 new():
         Bad size value.
                 Value: %s
         From Class::Utils:
                 Unknown parameter '%s'.

 run():
         Cannot create video.

=head1 EXAMPLE

 use strict;
 use warnings;

 use App::Video::Generator;

 # Run.
 App::Video::Generator->new->run;

 # Output like:
 # Usage: /tmp/7b3GEofrss [-d duration] [-f fps] [-h]
 #         [-i input_dir] [-s size] [-v] [--version] output_file
 #
 #         -d duration     Duration in numeric value or with ms/s/min/h suffix
 #                         (default value is 10000 [=10s]).
 #         -f fps          Frame rate
 #         -h              Print help.
 #         -i input_dir    Input directory with images (default value is nothing).
 #         -s size         Size (default value is 1920x1080).
 #         -v              Verbose mode.
 #         --version       Print version.

=head1 DEPENDENCIES

L<English>,
L<Error::Pure>,
L<Getopt::Std>,
L<Image::Select>,
L<Readonly>,
L<Video::Generator>.

=head1 SEE ALSO

=over

=item L<App::Image::Generator>

Perl class for image-generator application.

=back

=head1 REPOSITORY

L<https://github.com/michal-josef-spacek/App-Video-Generator>.

=head1 AUTHOR

Michal Josef Špaček L<mailto:skim@cpan.org>

L<http://skim.cz>

=head1 LICENSE AND COPYRIGHT

© 2015-2021 Michal Josef Špaček

BSD 2-Clause License

=head1 VERSION

0.10

=cut


Makefile.PL
use lib '.';
use strict;
use warnings;

use inc::Module::Install;

# Definition.
abstract 'Script to generate video.';
author 'Michal Josef Spacek <skim@cpan.org>';
author_requires 'English' => 0;
author_requires 'File::Object' => 0.05;
author_requires 'Test::Pod' => 0;
author_requires 'Test::Pod::Coverage' => 0;
install_script 'bin/video-generator';
license 'bsd';
name 'App-Video-Generator';
readme_from 'bin/video-generator';
recursive_author_tests('xt');
requires 'English' => 0;
requires 'Error::Pure' => 0.16;
requires 'Getopt::Std' => 0;
requires 'Image::Select' => 0.03;
requires 'Readonly' => 0;
requires 'Video::Generator' => 0.06;
requires 'perl' => '5.6.0';
resources 'bugtracker' => 'https://github.com/michal-josef-spacek/App-Video-Generator/issues';
resources 'homepage' => 'https://github.com/michal-josef-spacek/App-Video-Generator';
resources 'repository' => 'git://github.com/michal-josef-spacek/App-Video-Generator';
test_requires 'Test::More' => 0;
test_requires 'Test::NoWarnings' => 0;
tests_recursive;
version '0.10';

# Run.
WriteAll();
